File: client/lib/api/event-types.ts
/**
 * API client for event types management
 */

const API_BASE_URL =
  process.env.NEXT_PUBLIC_API_BASE_URL || "http://localhost:5000"

// Event Type interface
export interface EventType {
  id: string
  name: string
  description?: string | null
  duration: number
  color: string
  category: "interview" | "counseling" | "assessment" | "exit" | "custom"
  locationType: "in_person" | "video" | "phone"
  locationDetails?: string | null
  isActive: boolean
  requiresApproval: boolean
  maxBookingsPerDay?: number | null
  bufferBefore: number
  bufferAfter: number
  scheduleId?: string | null
  scheduleName?: string | null
  createdAt: string
  updatedAt: string
}

// Input type for creating/updating
export interface EventTypeInput {
  name: string
  description?: string | null
  duration?: number
  color?: string
  category?: "interview" | "counseling" | "assessment" | "exit" | "custom"
  locationType?: "in_person" | "video" | "phone"
  locationDetails?: string | null
  isActive?: boolean
  requiresApproval?: boolean
  maxBookingsPerDay?: number | null
  bufferBefore?: number
  bufferAfter?: number
  scheduleId?: string | null
}

/**
 * Get auth token from Supabase session
 */
async function getAuthToken(): Promise<string | null> {
  try {
    const { createClient } = await import("@/lib/client")
    const supabase = createClient()
    const {
      data: { session },
    } = await supabase.auth.getSession()
    return session?.access_token || null
  } catch (error) {
    console.error("Error getting auth token:", error)
    return null
  }
}

/**
 * Make authenticated API request
 */
async function apiRequest<T>(
  endpoint: string,
  options: RequestInit = {}
): Promise<T> {
  const token = await getAuthToken()

  if (!token) {
    throw new Error("Not authenticated")
  }

  const response = await fetch(`${API_BASE_URL}${endpoint}`, {
    ...options,
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${token}`,
      ...options.headers,
    },
  })

  if (!response.ok) {
    const error = await response
      .json()
      .catch(() => ({ error: "Unknown error" }))
    throw new Error(error.error || `API error: ${response.statusText}`)
  }

  return response.json()
}

/**
 * Get all event types for the current counselor
 */
export async function getEventTypes(): Promise<EventType[]> {
  try {
    const data = await apiRequest<{ eventTypes: EventType[] }>(
      "/api/event-types"
    )
    return data.eventTypes
  } catch (error) {
    console.error("Error fetching event types:", error)
    return []
  }
}

/**
 * Create a new event type
 */
export async function createEventType(
  input: EventTypeInput
): Promise<{ success: boolean; eventType?: EventType; error?: string }> {
  try {
    const result = await apiRequest<{ message: string; eventType: EventType }>(
      "/api/event-types",
      {
        method: "POST",
        body: JSON.stringify(input),
      }
    )
    return { success: true, eventType: result.eventType }
  } catch (error) {
    console.error("Error creating event type:", error)
    return {
      success: false,
      error:
        error instanceof Error ? error.message : "Failed to create event type",
    }
  }
}

/**
 * Update an event type
 */
export async function updateEventType(
  eventTypeId: string,
  input: Partial<EventTypeInput>
): Promise<{ success: boolean; eventType?: EventType; error?: string }> {
  try {
    const result = await apiRequest<{ message: string; eventType: EventType }>(
      `/api/event-types/${eventTypeId}`,
      {
        method: "PUT",
        body: JSON.stringify(input),
      }
    )
    return { success: true, eventType: result.eventType }
  } catch (error) {
    console.error("Error updating event type:", error)
    return {
      success: false,
      error:
        error instanceof Error ? error.message : "Failed to update event type",
    }
  }
}

/**
 * Delete an event type
 */
export async function deleteEventType(
  eventTypeId: string
): Promise<{ success: boolean; error?: string }> {
  try {
    await apiRequest(`/api/event-types/${eventTypeId}`, {
      method: "DELETE",
    })
    return { success: true }
  } catch (error) {
    console.error("Error deleting event type:", error)
    return {
      success: false,
      error:
        error instanceof Error ? error.message : "Failed to delete event type",
    }
  }
}

/**
 * Duplicate an event type
 */
export async function duplicateEventType(
  eventTypeId: string,
  newName: string
): Promise<{ success: boolean; eventType?: EventType; error?: string }> {
  try {
    const result = await apiRequest<{ message: string; eventType: EventType }>(
      `/api/event-types/${eventTypeId}/duplicate`,
      {
        method: "POST",
        body: JSON.stringify({ name: newName }),
      }
    )
    return { success: true, eventType: result.eventType }
  } catch (error) {
    console.error("Error duplicating event type:", error)
    return {
      success: false,
      error:
        error instanceof Error
          ? error.message
          : "Failed to duplicate event type",
    }
  }
}

/**
 * Get public event types for a counselor (for student booking)
 */
export async function getPublicEventTypes(
  counselorId: string,
  category?: "interview" | "counseling" | "assessment" | "exit" | "custom"
): Promise<EventType[]> {
  try {
    let url = `${API_BASE_URL}/api/event-types/public/${counselorId}`
    if (category) {
      url += `?category=${category}`
    }
    const response = await fetch(url)

    if (!response.ok) {
      throw new Error("Failed to fetch event types")
    }

    const data = await response.json()
    return data.eventTypes
  } catch (error) {
    console.error("Error fetching public event types:", error)
    return []
  }
}

// Simplified event type for schedule linking display
export interface LinkedEventType {
  id: string
  name: string
  color: string
  isActive: boolean
}

/**
 * Get event types linked to a specific schedule
 */
export async function getEventTypesBySchedule(
  scheduleId: string
): Promise<LinkedEventType[]> {
  try {
    const data = await apiRequest<{ eventTypes: LinkedEventType[] }>(
      `/api/event-types/by-schedule/${scheduleId}`
    )
    return data.eventTypes
  } catch (error) {
    console.error("Error fetching event types by schedule:", error)
    return []
  }
}

// Predefined color options for event types
export const EVENT_TYPE_COLORS = [
  { name: "Maroon", value: "#991b1b" },
  { name: "Red", value: "#dc2626" },
  { name: "Orange", value: "#ea580c" },
  { name: "Amber", value: "#d97706" },
  { name: "Yellow", value: "#ca8a04" },
  { name: "Lime", value: "#65a30d" },
  { name: "Green", value: "#16a34a" },
  { name: "Emerald", value: "#059669" },
  { name: "Teal", value: "#0d9488" },
  { name: "Cyan", value: "#0891b2" },
  { name: "Sky", value: "#0284c7" },
  { name: "Blue", value: "#2563eb" },
  { name: "Indigo", value: "#4f46e5" },
  { name: "Violet", value: "#7c3aed" },
  { name: "Purple", value: "#9333ea" },
  { name: "Fuchsia", value: "#c026d3" },
  { name: "Pink", value: "#db2777" },
  { name: "Rose", value: "#e11d48" },
]

// Duration options in minutes
export const DURATION_OPTIONS = [
  { label: "15 minutes", value: 15 },
  { label: "30 minutes", value: 30 },
  { label: "45 minutes", value: 45 },
  { label: "1 hour", value: 60 },
  { label: "1.5 hours", value: 90 },
  { label: "2 hours", value: 120 },
]

// Buffer time options in minutes
export const BUFFER_OPTIONS = [
  { label: "No buffer", value: 0 },
  { label: "5 minutes", value: 5 },
  { label: "10 minutes", value: 10 },
  { label: "15 minutes", value: 15 },
  { label: "30 minutes", value: 30 },
]

// Location type options
export const LOCATION_TYPE_OPTIONS = [
  { label: "In Person", value: "in_person" as const },
  { label: "Video Call", value: "video" as const },
  { label: "Phone Call", value: "phone" as const },
]

export const CATEGORY_OPTIONS = [
  {
    label: "Interview",
    value: "interview" as const,
    description: "Initial interview and assessment",
  },
  {
    label: "Counseling",
    value: "counseling" as const,
    description: "Regular counseling session",
  },
  {
    label: "Assessment",
    value: "assessment" as const,
    description: "Psychological tests and assessments",
  },
  {
    label: "Exit Interview",
    value: "exit" as const,
    description: "Final session and clearance",
  },
  {
    label: "Custom",
    value: "custom" as const,
    description: "Custom appointment type",
  },
]

File: server/app/routes/event_types.py
"""
Event Types API routes for counselor appointment type management
"""

from flask import Blueprint, request, jsonify
from app.utils.auth import require_auth
from app.services.supabase_service import get_supabase_client

event_types_bp = Blueprint("event_types", __name__, url_prefix="/api/event-types")


@event_types_bp.route("", methods=["GET"])
@require_auth
def get_event_types(user_id: str):
    """Get all event types for the current counselor"""
    try:
        supabase = get_supabase_client(use_service_role=True)
        
        response = (
            supabase.table("event_types")
            .select("*, counselor_schedules(id, name)")
            .eq("counselor_id", user_id)
            .order("created_at", desc=False)
            .execute()
        )
        
        event_types = []
        for et in response.data:
            schedule = et.get("counselor_schedules")
            event_types.append({
                "id": et["id"],
                "name": et["name"],
                "description": et.get("description"),
                "duration": et["duration"],
                "color": et["color"],
                "category": et.get("category", "counseling"),
                "locationType": et["location_type"],
                "locationDetails": et.get("location_details"),
                "isActive": et["is_active"],
                "requiresApproval": et["requires_approval"],
                "maxBookingsPerDay": et.get("max_bookings_per_day"),
                "bufferBefore": et["buffer_before"],
                "bufferAfter": et["buffer_after"],
                "scheduleId": et.get("schedule_id"),
                "scheduleName": schedule["name"] if schedule else None,
                "createdAt": et["created_at"],
                "updatedAt": et["updated_at"],
            })
        
        return jsonify({"eventTypes": event_types}), 200
        
    except Exception as e:
        return jsonify({"error": str(e)}), 500


@event_types_bp.route("", methods=["POST"])
@require_auth
def create_event_type(user_id: str):
    """Create a new event type"""
    try:
        data = request.get_json()
        name = data.get("name", "").strip()
        
        if not name:
            return jsonify({"error": "Event type name is required"}), 400
        
        supabase = get_supabase_client(use_service_role=True)
        
        # Build insert data
        insert_data = {
            "counselor_id": user_id,
            "name": name,
            "description": data.get("description"),
            "duration": data.get("duration", 30),
            "color": data.get("color", "#991b1b"),
            "category": data.get("category", "counseling"),
            "location_type": data.get("locationType", "in_person"),
            "location_details": data.get("locationDetails"),
            "is_active": data.get("isActive", True),
            "requires_approval": data.get("requiresApproval", False),
            "max_bookings_per_day": data.get("maxBookingsPerDay"),
            "buffer_before": data.get("bufferBefore", 0),
            "buffer_after": data.get("bufferAfter", 0),
            "schedule_id": data.get("scheduleId"),
        }
        
        response = supabase.table("event_types").insert(insert_data).execute()
        
        if not response.data:
            return jsonify({"error": "Failed to create event type"}), 500
        
        new_event_type = response.data[0]
        
        # Fetch schedule name if schedule_id exists
        schedule_name = None
        if new_event_type.get("schedule_id"):
            schedule_response = (
                supabase.table("counselor_schedules")
                .select("name")
                .eq("id", new_event_type["schedule_id"])
                .execute()
            )
            if schedule_response.data:
                schedule_name = schedule_response.data[0]["name"]
        
        return jsonify({
            "message": "Event type created successfully",
            "eventType": {
                "id": new_event_type["id"],
                "name": new_event_type["name"],
                "description": new_event_type.get("description"),
                "duration": new_event_type["duration"],
                "color": new_event_type["color"],
                "category": new_event_type.get("category", "counseling"),
                "locationType": new_event_type["location_type"],
                "locationDetails": new_event_type.get("location_details"),
                "isActive": new_event_type["is_active"],
                "requiresApproval": new_event_type["requires_approval"],
                "maxBookingsPerDay": new_event_type.get("max_bookings_per_day"),
                "bufferBefore": new_event_type["buffer_before"],
                "bufferAfter": new_event_type["buffer_after"],
                "scheduleId": new_event_type.get("schedule_id"),
                "scheduleName": schedule_name,
                "createdAt": new_event_type["created_at"],
                "updatedAt": new_event_type["updated_at"],
            }
        }), 201
        
    except Exception as e:
        return jsonify({"error": str(e)}), 500


@event_types_bp.route("/<string:event_type_id>", methods=["PUT"])
@require_auth
def update_event_type(user_id: str, event_type_id: str):
    """Update an event type"""
    try:
        data = request.get_json()
        supabase = get_supabase_client(use_service_role=True)
        
        # Verify ownership
        current = (
            supabase.table("event_types")
            .select("*")
            .eq("id", event_type_id)
            .eq("counselor_id", user_id)
            .execute()
        )
        
        if not current.data:
            return jsonify({"error": "Event type not found"}), 404
        
        # Build update data (only include fields that are provided)
        update_data = {}
        
        if "name" in data:
            name = data["name"].strip()
            if not name:
                return jsonify({"error": "Event type name cannot be empty"}), 400
            update_data["name"] = name
        
        if "description" in data:
            update_data["description"] = data["description"]
        
        if "duration" in data:
            update_data["duration"] = data["duration"]
        
        if "color" in data:
            update_data["color"] = data["color"]
        
        if "locationType" in data:
            update_data["location_type"] = data["locationType"]
        
        if "locationDetails" in data:
            update_data["location_details"] = data["locationDetails"]
        
        if "isActive" in data:
            update_data["is_active"] = data["isActive"]
        
        if "requiresApproval" in data:
            update_data["requires_approval"] = data["requiresApproval"]
        
        if "maxBookingsPerDay" in data:
            update_data["max_bookings_per_day"] = data["maxBookingsPerDay"]
        
        if "bufferBefore" in data:
            update_data["buffer_before"] = data["bufferBefore"]
        
        if "bufferAfter" in data:
            update_data["buffer_after"] = data["bufferAfter"]
        
        if "category" in data:
            update_data["category"] = data["category"]
        
        if "scheduleId" in data:
            update_data["schedule_id"] = data["scheduleId"]
        
        if not update_data:
            return jsonify({"error": "No changes provided"}), 400
        
        response = (
            supabase.table("event_types")
            .update(update_data)
            .eq("id", event_type_id)
            .eq("counselor_id", user_id)
            .execute()
        )
        
        if not response.data:
            return jsonify({"error": "Failed to update event type"}), 500
        
        updated = response.data[0]
        
        # Fetch schedule name if schedule_id exists
        schedule_name = None
        if updated.get("schedule_id"):
            schedule_response = (
                supabase.table("counselor_schedules")
                .select("name")
                .eq("id", updated["schedule_id"])
                .execute()
            )
            if schedule_response.data:
                schedule_name = schedule_response.data[0]["name"]
        
        return jsonify({
            "message": "Event type updated successfully",
            "eventType": {
                "id": updated["id"],
                "name": updated["name"],
                "description": updated.get("description"),
                "duration": updated["duration"],
                "color": updated["color"],
                "category": updated.get("category", "counseling"),
                "locationType": updated["location_type"],
                "locationDetails": updated.get("location_details"),
                "isActive": updated["is_active"],
                "requiresApproval": updated["requires_approval"],
                "maxBookingsPerDay": updated.get("max_bookings_per_day"),
                "bufferBefore": updated["buffer_before"],
                "bufferAfter": updated["buffer_after"],
                "scheduleId": updated.get("schedule_id"),
                "scheduleName": schedule_name,
                "createdAt": updated["created_at"],
                "updatedAt": updated["updated_at"],
            }
        }), 200
        
    except Exception as e:
        return jsonify({"error": str(e)}), 500


@event_types_bp.route("/<string:event_type_id>", methods=["DELETE"])
@require_auth
def delete_event_type(user_id: str, event_type_id: str):
    """Delete an event type"""
    try:
        supabase = get_supabase_client(use_service_role=True)
        
        # Verify ownership
        current = (
            supabase.table("event_types")
            .select("id")
            .eq("id", event_type_id)
            .eq("counselor_id", user_id)
            .execute()
        )
        
        if not current.data:
            return jsonify({"error": "Event type not found"}), 404
        
        # Delete the event type
        supabase.table("event_types").delete().eq(
            "id", event_type_id
        ).eq("counselor_id", user_id).execute()
        
        return jsonify({"message": "Event type deleted successfully"}), 200
        
    except Exception as e:
        return jsonify({"error": str(e)}), 500


@event_types_bp.route("/<string:event_type_id>/duplicate", methods=["POST"])
@require_auth
def duplicate_event_type(user_id: str, event_type_id: str):
    """Duplicate an event type"""
    try:
        data = request.get_json()
        new_name = data.get("name", "").strip()
        
        if not new_name:
            return jsonify({"error": "New event type name is required"}), 400
        
        supabase = get_supabase_client(use_service_role=True)
        
        # Get source event type
        source = (
            supabase.table("event_types")
            .select("*")
            .eq("id", event_type_id)
            .eq("counselor_id", user_id)
            .execute()
        )
        
        if not source.data:
            return jsonify({"error": "Source event type not found"}), 404
        
        source_et = source.data[0]
        
        # Create duplicate
        insert_data = {
            "counselor_id": user_id,
            "name": new_name,
            "description": source_et.get("description"),
            "duration": source_et["duration"],
            "color": source_et["color"],
            "category": source_et.get("category", "counseling"),
            "location_type": source_et["location_type"],
            "location_details": source_et.get("location_details"),
            "is_active": False,  # Start duplicates as inactive
            "requires_approval": source_et["requires_approval"],
            "max_bookings_per_day": source_et.get("max_bookings_per_day"),
            "buffer_before": source_et["buffer_before"],
            "buffer_after": source_et["buffer_after"],
            "schedule_id": source_et.get("schedule_id"),
        }
        
        response = supabase.table("event_types").insert(insert_data).execute()
        
        if not response.data:
            return jsonify({"error": "Failed to duplicate event type"}), 500
        
        new_event_type = response.data[0]
        
        # Fetch schedule name if schedule_id exists
        schedule_name = None
        if new_event_type.get("schedule_id"):
            schedule_response = (
                supabase.table("counselor_schedules")
                .select("name")
                .eq("id", new_event_type["schedule_id"])
                .execute()
            )
            if schedule_response.data:
                schedule_name = schedule_response.data[0]["name"]
        
        return jsonify({
            "message": "Event type duplicated successfully",
            "eventType": {
                "id": new_event_type["id"],
                "name": new_event_type["name"],
                "description": new_event_type.get("description"),
                "duration": new_event_type["duration"],
                "color": new_event_type["color"],
                "category": new_event_type.get("category", "counseling"),
                "locationType": new_event_type["location_type"],
                "locationDetails": new_event_type.get("location_details"),
                "isActive": new_event_type["is_active"],
                "requiresApproval": new_event_type["requires_approval"],
                "maxBookingsPerDay": new_event_type.get("max_bookings_per_day"),
                "bufferBefore": new_event_type["buffer_before"],
                "bufferAfter": new_event_type["buffer_after"],
                "scheduleId": new_event_type.get("schedule_id"),
                "scheduleName": schedule_name,
                "createdAt": new_event_type["created_at"],
                "updatedAt": new_event_type["updated_at"],
            }
        }), 201
        
    except Exception as e:
        return jsonify({"error": str(e)}), 500


@event_types_bp.route("/by-schedule/<string:schedule_id>", methods=["GET"])
@require_auth
def get_event_types_by_schedule(user_id: str, schedule_id: str):
    """Get event types linked to a specific schedule"""
    try:
        supabase = get_supabase_client(use_service_role=True)
        
        # Get event types that use this schedule
        response = (
            supabase.table("event_types")
            .select("id, name, color, is_active")
            .eq("counselor_id", user_id)
            .eq("schedule_id", schedule_id)
            .eq("is_active", True)
            .order("name")
            .execute()
        )
        
        event_types = [
            {
                "id": et["id"],
                "name": et["name"],
                "color": et["color"],
                "isActive": et["is_active"],
            }
            for et in response.data
        ]
        
        return jsonify({"eventTypes": event_types}), 200
        
    except Exception as e:
        return jsonify({"error": str(e)}), 500


@event_types_bp.route("/public/<string:counselor_id>", methods=["GET"])
def get_public_event_types(counselor_id: str):
    """Get active event types for a counselor (public endpoint for students)"""
    try:
        supabase = get_supabase_client(use_service_role=True)
        
        # Optional category filter
        category = request.args.get("category")
        
        query = (
            supabase.table("event_types")
            .select("id, name, description, duration, color, category, location_type, location_details")
            .eq("counselor_id", counselor_id)
            .eq("is_active", True)
        )
        
        if category:
            query = query.eq("category", category)
        
        response = query.order("name").execute()
        
        event_types = []
        for et in response.data:
            event_types.append({
                "id": et["id"],
                "name": et["name"],
                "description": et.get("description"),
                "duration": et["duration"],
                "color": et["color"],
                "category": et.get("category", "counseling"),
                "locationType": et["location_type"],
                "locationDetails": et.get("location_details"),
            })
        
        return jsonify({"eventTypes": event_types}), 200
        
    except Exception as e:
        return jsonify({"error": str(e)}), 500
